### PR TITLE
fix stuck when publish to full queue

### PR DESCRIPTION
**Pull request by Infactum**
possible fix for #215

### CODE REVIEW

File: src/Channel.cpp [hunk 1]
```diff
@@ -524,17 +524,25 @@ void Channel::BasicPublish(const std::string &exchange_name,
...
+  const boost::array<boost::uint32_t, 3> PUBLISH_ACK = {
+      {AMQP_BASIC_ACK_METHOD, AMQP_BASIC_RETURN_METHOD, AMQP_BASIC_NACK_METHOD}};
   amqp_frame_t response;
   boost::array<amqp_channel_t, 1> channels = {{channel}};
   m_impl->GetMethodOnChannel(channels, response, PUBLISH_ACK);
 
+  if (AMQP_BASIC_NACK_METHOD == response.payload.method.id) {
+    m_impl->ReturnChannel(channel);
+    m_impl->MaybeReleaseBuffersOnChannel(channel);
+    throw std::runtime_error("Queue is full");
```
Comment (Infactum) [inline src/Channel.cpp]: Not sure what type of exception to use here.

**Comment by grievertime** [comment]
Any news on this pull request? I'd like to not fork the entire project for just this little bug.

**Comment by alanxz** [comment]
> Any news on this pull request? I'd like to not fork the entire project for just this little bug.

This stalled as I didn't have time to work through the issues at the time.

I'm willing to get this feature added to SimpleAmqpClient.  The notable missing pieces from this are: 

1) Instead of `throw std::runtime_error`, a new exception type `AmqpClient::MessageRejected` should be created, and this function should throw that

2) Add a test that validates that this works as intended.

You'll probably want to open a new PR.

**Comment by Infactum** [comment]
Corrected exception type and added a test.

**Comment by alanxz** [comment]
Merged this in #248 with some minor tweaks to commit message and make it compatible with C++98 (yes, this library needs to be updated to something like C++17, just not there yet).

Thanks again.

**Pull request closed without merging**